File: frontend/src/features/research/numericDisplay.ts
const FIAT_UNITS = new Set([
  "$",
  "¥",
  "AUD",
  "CAD",
  "CHF",
  "CNY",
  "EUR",
  "GBP",
  "HKD",
  "JPY",
  "KRW",
  "USD",
  "円",
]);
const PERCENT_UNITS = new Set(["%", "PCT", "PERCENT"]);
const RATIO_UNITS = new Set(["X", "倍"]);

export function decisionFractionDigits(value: number, unit?: string): number {
  const normalized = (unit ?? "").trim().toUpperCase();
  const absolute = Math.abs(value);
  if (PERCENT_UNITS.has(normalized) || RATIO_UNITS.has(normalized)) return 2;
  if (absolute > 0 && absolute < 1) {
    const magnitude = Math.floor(Math.log10(absolute));
    return Math.min(8, Math.max(0, 3 - magnitude));
  }
  if (FIAT_UNITS.has(normalized)) return 2;
  return 4;
}

export function decisionNumberLocale(language?: string): string {
  const normalized = (language ?? "").toLowerCase();
  if (normalized.startsWith("zh")) return "zh-CN";
  if (normalized.startsWith("ja")) return "ja-JP";
  return "en-US";
}

export function formatDecisionNumber(
  value: number,
  unit?: string,
  language?: string,
): string {
  if (!Number.isFinite(value)) return String(value);
  const digits = decisionFractionDigits(value, unit);
  return new Intl.NumberFormat(decisionNumberLocale(language), {
    useGrouping: true,
    minimumFractionDigits: 0,
    maximumFractionDigits: digits,
  }).format(Object.is(value, -0) ? 0 : value);
}

File: tradingagents/domain/numeric_display.py
"""Deterministic, presentation-only formatting for audited decision numbers."""

from __future__ import annotations

import math
from decimal import ROUND_HALF_UP, Decimal

_FIAT_UNITS = {
    "$",
    "¥",
    "AUD",
    "CAD",
    "CHF",
    "CNY",
    "EUR",
    "GBP",
    "HKD",
    "JPY",
    "KRW",
    "USD",
    "円",
}
_PERCENT_UNITS = {"%", "PCT", "PERCENT"}
_RATIO_UNITS = {"X", "倍"}


def decision_fraction_digits(value: float, unit: str | None) -> int:
    """Return reading precision without changing the canonical numeric value."""

    normalized = (unit or "").strip().upper()
    absolute = abs(value)
    if normalized in _PERCENT_UNITS or normalized in _RATIO_UNITS:
        return 2
    if 0 < absolute < 1:
        magnitude = math.floor(math.log10(absolute))
        return min(8, max(0, 3 - magnitude))
    if normalized in _FIAT_UNITS:
        return 2
    return 4


def format_decision_number(
    value: float,
    unit: str | None = None,
    *,
    output_language: str | None = None,
) -> str:
    """Format a finite value for Web/Markdown while retaining raw audit data."""

    del output_language  # en, zh-CN, and ja share stable decimal/group separators.
    if not math.isfinite(value):
        return str(value)
    digits = decision_fraction_digits(value, unit)
    quantum = Decimal(1).scaleb(-digits)
    rounded = Decimal(str(value)).quantize(quantum, rounding=ROUND_HALF_UP)
    if rounded == 0:
        rounded = abs(rounded)
    rendered = f"{rounded:,.{digits}f}"
    if "." in rendered:
        rendered = rendered.rstrip("0").rstrip(".")
    return rendered
